ta/pkcs11/src/processing_symm.c

```diff
@@ -833,3 +833,42 @@ enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
 
 	return rc;
 }
+
+enum pkcs11_rc wrap_data_by_symm_enc(struct pkcs11_session *session,
+				     void *data, uint32_t data_sz,
+				     void **out_buf, uint32_t *out_sz)
+
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct active_processing *proc = session->processing;
+	void *in_buf = NULL;
+	uint32_t align = 0;
+	uint32_t in_sz = 0;
+
+	switch (proc->mecha_type) {
+	case PKCS11_CKM_AES_ECB:
+	case PKCS11_CKM_AES_CBC:
+		align = data_sz % TEE_AES_BLOCK_SIZE;
+		if (align) {
+			in_sz = data_sz + (TEE_AES_BLOCK_SIZE - align);
+			in_buf = TEE_Malloc(in_sz, TEE_MALLOC_FILL_ZERO);
```

Comment:
If data buffer is big, we may prefer allocate only the last block.
We could address this here or in a later change.

Reply:
I can change this to - 
                align = data_sz % TEE_AES_BLOCK_SIZE;
                if (align) {
                        if (data_sz > TEE_AES_BLOCK_SIZE) {
                                in_sz = data_sz - align;
                                res = TEE_CipherUpdate(proc->tee_op_handle,
                                                       data, in_sz, *out_buf,
                                                       out_sz);
                                if (res)
                                        return tee2pkcs_error(res);
                        } 
                        in_buf = TEE_Malloc(TEE_AES_BLOCK_SIZE,
                                                        TEE_MALLOC_FILL_ZERO);
                        if (!in_buf)
                                return PKCS11_CKR_ARGUMENTS_BAD;

                        TEE_MemMove(in_buf, (uint8_t *)data + in_sz, align);
                        in_sz = TEE_AES_BLOCK_SIZE;
                } else {
                        in_buf = data;
                        in_sz = data_sz;
                }

                res = TEE_CipherDoFinal(proc->tee_op_handle, in_buf, in_sz,
                                        *out_buf, out_sz);
                if (align)
                        TEE_Free(in_buf);

Reply:
> > Tested out CKM_AES_CBC and seems to work.
> > Just for fun -- will try circle between OP-TEE and SoftHSM.
> 
> SoftHSMv2 seems to be a bit conservative in here.
> 
> Unwrap for devel branch (eg. not yet in release):
> 
> * wrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6458
>   
>   * CKM_AES_KEY_WRAP -- if detected in crypto backend
>   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
>   * CKM_RSA_PKCS
>   * CKM_RSA_PKCS_OAEP
>   * CKM_AES_CBC
>   * CKM_AES_CBC_PAD
> * unwrap: https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/SoftHSM.cpp#L6725
>   
>   * CKM_AES_KEY_WRAP -- if detected in crypto backend
>   * CKM_AES_KEY_WRAP_PAD -- if detected in crypto backend
>   * CKM_AES_CBC_PAD
>   * CKM_DES3_CBC_PAD
> 
> Kinda understand that DES3 is not in wrapping part.
> 
> RSA_PKCS and RSA_PKCS_OAEP I understand for wrapping but don't understand why it is not in unwrap.
> 
> For one commercial HSM in order to export AES key at all it needs to be explicitly enabled. And after enabled it barks in the log that now it was accessed.
> 
> Now I am wondering should we have mk/config.mk options to enable these weaker exports?

Sorry, I didn't exactly understand what you mean here ? Can you explain more.

Reply:
You somehow replied it in here -- I try to clarify it in "base comment"